src/lib/gadgets/foreign-field.ts

```diff
@@ -92,12 +107,230 @@ function singleAdd(x: Field3, y: Field3, sign: Sign, f: bigint) {
   return { result: [r0, r1, r2] satisfies Field3, overflow };
 }
 
+function multiply(a: Field3, b: Field3, f: bigint): Field3 {
+  assert(f < 1n << 259n, 'Foreign modulus fits in 259 bits');
+
+  // constant case
+  if (a.every((x) => x.isConstant()) && b.every((x) => x.isConstant())) {
+    let ab = Field3.toBigint(a) * Field3.toBigint(b);
+    return Field3.from(mod(ab, f));
+  }
+
+  // provable case
+  let { r01, r2, q, q2Bound } = multiplyNoRangeCheck(a, b, f);
+
+  // limb range checks on quotient and remainder
+  multiRangeCheck(q);
+  let r = compactMultiRangeCheck(r01, r2);
+
+  // range check on q and r bounds
+  // TODO: this uses one RC too many.. need global RC stack, or get rid of bounds checks
+  let r2Bound = weakBound(r2, f);
+  multiRangeCheck([q2Bound, r2Bound, Field.from(0n)]);
+
+  return r;
+}
+
+function inverse(x: Field3, f: bigint): Field3 {
+  assert(f < 1n << 259n, 'Foreign modulus fits in 259 bits');
+
+  // constant case
+  if (x.every((x) => x.isConstant())) {
+    let xInv = modInverse(Field3.toBigint(x), f);
+    assert(xInv !== undefined, 'inverse exists');
+    return Field3.from(xInv);
+  }
+
+  // provable case
+  let xInv = exists(3, () => {
+    let xInv = modInverse(Field3.toBigint(x), f);
+    return xInv === undefined ? [0n, 0n, 0n] : split(xInv);
+  });
+  multiRangeCheck(xInv);
+  let xInv2Bound = weakBound(xInv[2], f);
+
+  let one: Field2 = [Field.from(1n), Field.from(0n)];
+  let q2Bound = assertMul(x, xInv, one, f);
+
+  // range check on q and result bounds
+  // TODO: this uses one RC too many.. need global RC stack
+  multiRangeCheck([q2Bound, xInv2Bound, Field.from(0n)]);
+
+  return xInv;
+}
+
+function divide(
+  x: Field3,
+  y: Field3,
+  f: bigint,
+  { allowZeroOverZero = false } = {}
+) {
+  assert(f < 1n << 259n, 'Foreign modulus fits in 259 bits');
+
+  // constant case
+  if (x.every((x) => x.isConstant()) && y.every((x) => x.isConstant())) {
+    let yInv = modInverse(Field3.toBigint(y), f);
+    assert(yInv !== undefined, 'inverse exists');
+    return Field3.from(mod(Field3.toBigint(x) * yInv, f));
+  }
+
+  // provable case
+  // to show that z = x/y, we prove that z*y = x and y != 0 (the latter avoids the unconstrained 0/0 case)
+  let z = exists(3, () => {
+    let yInv = modInverse(Field3.toBigint(y), f);
+    if (yInv === undefined) return [0n, 0n, 0n];
+    return split(mod(Field3.toBigint(x) * yInv, f));
+  });
+  multiRangeCheck(z);
+  let z2Bound = weakBound(z[2], f);
+  let q2Bound = assertMul(z, y, x, f);
+
+  // range check on q and result bounds
+  multiRangeCheck([q2Bound, z2Bound, Field.from(0n)]);
+
+  if (!allowZeroOverZero) {
+    // assert that y != 0 mod f by checking that it doesn't equal 0 or f
+    // this works because we assume y[2] <= f2
+    // TODO is this the most efficient way?
```

Comment:
What do you mean with proving y has an inverse? Do you mean prove that it has an inverse in the foreign field? That would be another ff multiplication, which is much more expensive than what I'm doing here (that's the reason that my `div()` is not just `inv()` + `mul()`)

Or do you mean proving that y has an inverse in the native field? That's not possible, because y is a vector of 3 limbs. If we would sum them and prove that the result has an inverse, this would only prove that y is not 0. But what if y is f. Also, this would ruin completeness, since it would not allow y to equal the native modulus (or any multiple of it), but that should actually be allowed

Reply:
The approach here is to check that y does not equal 0, and it does not equal f, by proving that not all the individual limbs are equal to 0 or f

Reply:
it uses the `equals()` gadget for native field elements, which you can check the code for here: https://github.com/o1-labs/o1js/blob/aec1e893774714696e621a64bb666d2c1d8f57ca/src/lib/field.ts#L651-L656

Reply:
the real magic happens inside `isZero()`:
https://github.com/o1-labs/o1js/blob/aec1e893774714696e621a64bb666d2c1d8f57ca/src/lib/field.ts#L609-L629

Reply:
> and it does not equal f, by proving that not all the individual limbs

What if it was any other multiple of `f`? You don't consider this case because it is assumed to be range-checked to be at most `f`?

Reply:
>     `y01.equals(0n).and(y[2].equals(0n)).assertFalse();`

I didn't know you could do this type of assert_neq's in o1js. How does it work under the hood though? In Kimchi, we just create the constraints to satisfy the properties, not the opposite. Meaning, if one wanted to check that a value is **not** zero in one gate, one would have to either write all permitted values (insanely complex) or provide the inverse in the witness and just prove that $1 - x*inv = 0$.